### PR TITLE
PullRequest for Issue18 and Issue19: implement client to support multiple named server

### PR DESCRIPTION
**Pull request by XuetaoLiu-S**
- Add AMDSClientAppController to generalize the client side usage of AMDS and manage multiple connected named server;
- Add AMDSServer to manage the connected named server and active continuous connection
- rework on AMDSClient (AMDSClientUi), to simulate a typical usage of AMDSClientAppController
- some code structure to make the codes organized
- removed all the compile warnings
  https://github.com/acquaman/AcquamanDataServer/issues/18
  https://github.com/acquaman/AcquamanDataServer/issues/19

### CODE REVIEW

File: source/AMDSClientUi.cpp [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+}
+
+AMDSClientUi::~AMDSClientUi()
+{
+	disconnect(clientAppController_, SIGNAL(networkSessionOpening()), this, SLOT(onNetworkSessionOpening()));
+	disconnect(clientAppController_, SIGNAL(networkSessionOpened()), this, SLOT(onNetworkSessionOpened()));
+	disconnect(clientAppController_, SIGNAL(newServerConnected(QString)), this, SLOT(onNewServerConnected(QString)));
+	disconnect(clientAppController_, SIGNAL(serverError(int,QString,QString)), this, SLOT(onServerError(int,QString,QString)));
+	disconnect(clientAppController_, SIGNAL(requestDataReady(AMDSClientRequest*)), this, SLOT(onRequestDataReady(AMDSClientRequest*)));
+	disconnect(clientAppController_, SIGNAL(socketError(QString, QString)), this, SLOT(onSocketError(QString, QString)));
```
Comment (iainworkman) [inline source/AMDSClientUi.cpp]: Missing an int argument in this SIGNAL (should be socketError(int, QString, QString)).

File: source/AMDSClientUi.cpp [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+	activeContinuousConnectionComboBox_ = new QComboBox;
+	activeContinuousConnectionComboBox_->setFixedWidth(200);
+	activeContinuousConnectionComboBox_->addItem("");
+
+	resultsEdit_ = new QTextEdit();
+
+	connect(requestTypeComboBox_, SIGNAL(currentIndexChanged(QString)), this, SLOT(onRequestTypeChanged(QString)));
+
+
+//	amptekIndex = new QSpinBox();
```
Comment (dretrex) [inline source/AMDSClientUi.cpp:97]: Unnecessary commented out code?

Comment (XuetaoLiu-S) [inline source/AMDSClientUi.cpp:97]: This is from the original codes. I'm not sure about the usage or it yet. Will leave it there for a little while

File: source/AMDSClientUi.cpp [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+		}
+		break;
+	default:
+		AMDSErrorMon::alert(this, 0, QString("Invalide client request type: %1").arg(clientRequestType));
+	}
+}
+
+void AMDSClientUi::resetBufferListView(QStringList &bufferNames)
+{
+	QStringListModel *bufferNamesModel = new QStringListModel(this);
```
Comment (dretrex) [inline source/AMDSClientUi.cpp]: Does QListView handle the memory cleanup of this QStringListModel?  If not, you'll need to handle this as it is a memory leak.

File: source/AMDSClientUi.cpp (hunk 1 above)
Comment (XuetaoLiu-S) [inline source/AMDSClientUi.cpp]: Good catch! :8ball:

File: source/AMDSClientUi.h [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	void connectToServer();
+
+	/// ============= SLOTs to handle AMDSClientAppController signals =========
+	/// slot to handle the signal of networkSessionOpening
+	void onNetworkSessionOpening();
+	/// slot to handle the signal of networkSessionOpened
+	void onNetworkSessionOpened();
+
+	/// slot to handle the signal of newServerConnected (add the serverIdentifier to the combox and update the ui displays -- buffernames and active connections)
+	void onNewServerConnected(QString serverIdentifier);
```
Comment (dretrex) [inline source/AMDSClientUi.h]: All of the `QString` parameters should be `const QString &` and the `QStringList &` should be `const QStringList &`.

File: source/Connection/AMDSDataStream.cpp [hunk 5]
```diff
@@ -329,8 +329,8 @@ void AMDSDataStream::encodeClientRequestType(const AMDSClientRequest &clientRequ
 void AMDSDataStream::write(const AMDSClientRequest &clientRequest){
 	int errorCode = clientRequest.writeToDataStream(this);
 	if (errorCode != AMDS_CLIENTREQUEST_SUCCESS) {
-		QString errorMessage = AMDSClientRequestDefinitions::errorMessage(errorCode, AMDSClientRequestDefinitions::Write, clientRequest.requestType());
-		AMDSErrorMon::alert(0, errorCode, errorMessage);
+		QString errMessage = AMDSClientRequestDefinitions::errorMessage(errorCode, AMDSClientRequestDefinitions::Write, clientRequest.requestType());
```
Comment (dretrex) [inline source/Connection/AMDSDataStream.cpp]: I would prefer it if you did not make your names more terse.  `errorMessage` is preferred to `errMessage`.

File: source/AMDSClientUi.cpp [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+
+	activeServerComboBox_ = new QComboBox;
+	activeServerComboBox_->setFixedWidth(200);
+	activeServerComboBox_->addItem("");
+	connect(activeServerComboBox_, SIGNAL(currentIndexChanged(QString)), this, SLOT(onActiveServerChanged(QString)));
+
+	statusLabel_ = new QLabel(tr("This examples requires that you run the Acquaman Data Server example as well."));
+
+	connect(hostLineEdit_, SIGNAL(textChanged(QString)), this, SLOT(enableRequestDataButton()));
+	connect(portLineEdit_, SIGNAL(textChanged(QString)), this, SLOT(enableRequestDataButton()));
```
Comment (iainworkman) [inline source/AMDSClientUi.cpp]: There seems to be a little issue with these regarding the logic for setting the enabled state of the request data button.
When setting the IP or port to connect to initially there is no current text in the active server combo box (you've not connected yet), and thus the button is disabled. 
If you don't require a network session (as per line 69 of the app controller) none of the session signals will be fired and the button will only be re-enabled if you manually jump into either line edits after connecting and change a value.

File: source/Connection/AMDSServer.cpp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+void AMDSServer::setBufferNames(QStringList activeBufferNames) {
+	bufferNames_.clear();
+	bufferNames_.append(activeBufferNames);
+}
+
+AMDSClientTCPSocket * AMDSServer::establishSocketConnection()
+{
+	AMDSClientTCPSocket * clientTCPSocket = new AMDSClientTCPSocket(hostName_, portNumber_);
```
Comment (dretrex) [inline source/Connection/AMDSServer.cpp:31]: I'm pretty sure I've asked this and remember being satisfied with the answer, but just so my mind is at ease, is the memory of the created sockets here cleaned up in requested classes?

File: source/Connection/AMDSServer.cpp [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+	}
+
+	emit socketError(socketErrorCode, clientTCPSocket->socketKey(), errorString);
+	removeTCPSocket(clientTCPSocket);
+}
+
+void AMDSServer::removeTCPSocket(AMDSClientTCPSocket *clientTCPSocket)
+{
+	// try to remove from the active socket list
+	if (activeTCPSockets_.contains(clientTCPSocket->socketKey()))
```
Comment (dretrex) [inline source/Connection/AMDSServer.cpp:84]: And here it is.  Nevermind. :)

File: source/appController/AMDSClientAppController.cpp [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+		activeServers_.remove(key);
+	}
+
+	if (networkSession_)
+		networkSession_->deleteLater();
+}
+
+bool AMDSClientAppController::isSessionOpen()
+{
+	return (!networkSession_ || networkSession_->isOpen());
```
Comment (dretrex) [inline source/appController/AMDSClientAppController.cpp]: I believe this should actually be `return (networkSession_ && networkSession_->isOpen();`

File: source/appController/AMDSClientAppController.cpp [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+	return activeSocketKeys;
+}
+
+void AMDSClientAppController::openNetworkSession()
+{
+	if (!networkSession_) {
+		QNetworkConfigurationManager manager;
+		if (manager.capabilities() & QNetworkConfigurationManager::NetworkSessionRequired) {
+			// Get saved network configuration
+			QSettings settings(QSettings::UserScope, QLatin1String("Trolltech"));
```
Comment (dretrex) [inline source/appController/AMDSClientAppController.cpp]: Are the names TrollTech and QtNetwork required keywords?  If not, maybe make these more related to their Data Server purpose.

**Comment by iainworkman** [comment]
The code looks good. There was one issue that arose during testing though, and which I can reproduce each time on my laptop. 
If I run the server on my wlan0 interface, and then connect from the client on the loopback (127.0.0.1) I get a complete system crash. Mouse moves, nothing else responds - requires a hard power cycle of the machine to bring it back. From debugging this seems to occur somewhere in the server, not the client but I'm not 100%.
Time Estimate 1 hour

File: source/appController/AMDSClientAppController.h [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+public:
+	explicit AMDSClientAppController(QObject *parent = 0);
+	~AMDSClientAppController();
+
+	/// helper function to check whether the network session is openned or not
+	bool isSessionOpen();
+
+	/// helper function to return the server by given identifier
+	inline AMDSServer *getServerByServerIdentifier(QString serverIdentifier) { return activeServers_.value(serverIdentifier, 0);}
```
Comment (dretrex) [inline source/appController/AMDSClientAppController.h]: Here again, make sure that your class-based parameters use `const QString &` (or `const Class &` more generally).

**Comment by dretrex** [comment]
Time Estimate: 20 minutes

File: source/AMDSClientUi.cpp (hunk 6 above)
Comment (XuetaoLiu-CLS) [inline source/AMDSClientUi.cpp]: I will raise an issue for this and test is with your scenario

**Pull request merged**